lib/semantics/resolve-names.cc

```diff
@@ -1706,24 +1778,220 @@ void ResolveNamesVisitor::DeclareEntity(const parser::Name &name, Attrs attrs) {
+  //TODO: do something with derivedType
+  derivedTypeData_.reset();
+}
+bool DeclarationVisitor::Pre(const parser::DerivedTypeStmt &x) {
+  derivedTypeData_->name = std::get<parser::Name>(x.t).source;
+  BeginAttrs();
+  return true;
+}
+void DeclarationVisitor::Post(const parser::DerivedTypeStmt &x) {
+  derivedTypeData_->attrs = GetAttrs();
+  EndAttrs();
+}
+bool DeclarationVisitor::Pre(const parser::TypeAttrSpec::Extends &x) {
+  derivedTypeData_->extends = x.v.source;
+  return false;
+}
+bool DeclarationVisitor::Pre(const parser::PrivateStmt &x) {
+  derivedTypeData_->Private = true;
+  return false;
+}
+bool DeclarationVisitor::Pre(const parser::SequenceStmt &x) {
+  derivedTypeData_->sequence = true;
+  return false;
+}
+void DeclarationVisitor::Post(const parser::ComponentDecl &x) {
+  const auto &name = std::get<parser::Name>(x.t).source;
+  derivedTypeData_->dataComps.emplace_back(
+      *GetDeclTypeSpec(), name, GetAttrs(), arraySpec());
+  ClearArraySpec();
+}
+bool DeclarationVisitor::Pre(const parser::ProcedureDeclarationStmt &) {
+  CHECK(!interfaceName_);
+  return BeginDecl();
+}
+void DeclarationVisitor::Post(const parser::ProcedureDeclarationStmt &) {
+  interfaceName_ = nullptr;
+  EndDecl();
+}
+bool DeclarationVisitor::Pre(const parser::ProcComponentDefStmt &) {
+  CHECK(!interfaceName_);
+  return true;
+}
+void DeclarationVisitor::Post(const parser::ProcComponentDefStmt &) {
+  interfaceName_ = nullptr;
+}
+void DeclarationVisitor::Post(const parser::ProcInterface &x) {
+  if (auto *name = std::get_if<parser::Name>(&x.u)) {
+    interfaceName_ = &name->source;
+  }
+}
+
+void DeclarationVisitor::Post(const parser::ProcDecl &x) {
+  const auto &name = std::get<parser::Name>(x.t);
+  ProcInterface interface;
+  if (interfaceName_) {
+    auto *symbol = FindSymbol(*interfaceName_);
+    if (!symbol) {
+      Say(*interfaceName_, "Explicit interface '%s' not found"_err_en_US);
+    } else if (!symbol->HasExplicitInterface()) {
+      Say2(*interfaceName_,
+          "'%s' is not an abstract interface or a procedure with an explicit interface"_err_en_US,
+          symbol->name(), "Declaration of '%s'"_en_US);
     } else {
-      CHECK(!"unexpected kind");
+      interface = *symbol;
     }
+  } else if (auto &type = GetDeclTypeSpec()) {
+    interface = *type;
+  }
+  if (derivedTypeData_) {
+    derivedTypeData_->procComps.emplace_back(
+        ProcDecl{name.source}, GetAttrs(), std::move(interface));
   } else {
-    SayAlreadyDeclared(name.source, symbol);
+    DeclareProcEntity(name, GetAttrs(), std::move(interface));
+  }
+}
+
+bool DeclarationVisitor::Pre(const parser::FinalProcedureStmt &x) {
+  for (const parser::Name &name : x.v) {
+    derivedTypeData_->finalProcs.push_back(name.source);
+  }
+  return false;
+}
+
+void DeclarationVisitor::SetType(
+    const SourceName &name, Symbol &symbol, const DeclTypeSpec &type) {
+  if (auto *details = symbol.detailsIf<EntityDetails>()) {
```

Comment:
The flow here was a little tricky to grasp at first glance but I don't have a more easily readable paradigm to suggest -- using a flag seems no better, and splitting the function into a "get type if there is one" function and a second "set type or crash" function, calling both in a wrapper, seems worse.